cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -49,6 +50,59 @@ Status CheckRelCommon(const RelMessage& rel) {
   return Status::OK();
 }
 
+Result<fs::FileInfoVector> GetGlobFiles(const std::shared_ptr<fs::FileSystem>& filesystem,
+                                        std::string& path) {
+  fs::FileInfoVector results, temp;
+  fs::FileSelector selector;
+  std::string cur;
+  size_t i = 0;
+
+#if _WIN32
+  auto split_path = fs::internal::SplitAbstractPath(path, '\\');
+  ARROW_ASSIGN_OR_RAISE(auto file, filesystem->GetFileInfo(split_path[i++] + "\\"));
+#else
+  auto split_path = fs::internal::SplitAbstractPath(path, '/');
+  ARROW_ASSIGN_OR_RAISE(auto file, filesystem->GetFileInfo("/"));
+#endif
```

Comment:
Filesystems always work on `/` separated paths (regardless of the separated used behind the scenes).  So this method should assume the path is always `/` separated.  Since you are working with a filesystem you shouldn't need `#if _WIN32`.